dateutil/parser/isoparser.py

```diff
@@ -0,0 +1,424 @@
+        """
+        if not 0 < week < 54:
+            raise ValueError('Invalid week: {}'.format(week))
+
+        if not 0 < day < 8:     # Range is 1-7
+            raise ValueError('Invalid weekday: {}'.format(day))
+
+        # Get week 1 for the specific year:
+        jan_4 = date(year, 1, 4)   # Week 1 always has January 4th in it
+        week_1 = jan_4 - timedelta(days=jan_4.isocalendar()[2] - 1)
+
+        # Now add the specific number of weeks and days to get what we want
+        week_offset = (week - 1) * 7 + (day - 1)
+        return week_1 + timedelta(days=week_offset)
+
+    @classmethod
+    def _parse_isotime(cls, timestr):
+        len_str = len(timestr)
+        components = [0, 0, 0, 0, None]
+        pos = 0
+        comp = -1
+
+        if len(timestr) < 2:
+            raise ValueError('ISO time too short')
+
+        has_sep = len_str >= 3 and timestr[2:3] == cls._TIME_SEP
+
+        while pos < len_str and comp < 5:
+            comp += 1
+
+            if timestr[pos:pos + 1] in b'-+Z':
+                # Detect time zone boundary
+                components[-1] = cls._parse_tzstr(timestr[pos:])
+                pos = len_str
+                break
+
+            if comp < 3:
+                # Hour, minute, second
+                components[comp] = int(timestr[pos:pos + 2])
+                pos += 2
+                if (has_sep and pos < len_str and
+                        timestr[pos:pos + 1] == cls._TIME_SEP):
+                    pos += 1
+
+            if comp == 3:
+                # Microsecond
+                if timestr[pos:pos + 1] != cls._MICRO_SEP:
+                    continue
+
+                pos += 1
+                us_str = cls._MICROSECOND_END_REGEX.split(timestr[pos:pos + 6],
+                                                          1)[0]
+
+                components[comp] = int(us_str) * 10**(6 - len(us_str))
+                pos += len(us_str)
+
+        if pos < len_str:
+            raise ValueError('Unused components in ISO string')
+
+        if components[0] == 24:
+            # Standard supports 00:00 and 24:00 as representations of midnight
+            if any(component != 0 for component in components[1:4]):
+                raise ValueError('Hour may only be 24 at 24:00:00.000')
+            components[0] = 0
+
+        return components
+
+    @classmethod
+    def _parse_tzstr(cls, tzstr, zero_as_utc=True):
+        if tzstr == b'Z':
+            return tz.tzutc()
+
+        if len(tzstr) not in {3, 5, 6}:
+            raise ValueError('Time zone offset must be 1, 3, 5 or 6 characters')
+
+        if tzstr[0:1] == b'-':
+            mult = -1
+        elif tzstr[0:1] == b'+':
+            mult = 1
+        else:
+            raise ValueError('Time zone offset requires sign')
+
+        hours = int(tzstr[1:3])
+        if len(tzstr) == 3:
+            minutes = 0
+        else:
+            minutes = int(tzstr[(4 if tzstr[3:4] == cls._TIME_SEP else 3):])
```

Comment:
By this point you know that `len(tzstr)>3` so the `tzstr[3:4]` trick is unnecessary, can just do `tzstr[3]`.

The checks below are probably sufficient, but I would advocate explicitly checking here

```
elif len(tzstr) == 5:
    if not tzstr[-2:].isdigit(): --> raise
elif len(tzstr) == 6:
    if not tzstr[-2:].isdigit() or tzstr[3] != cls._TIME_SEP --> raise
```

Reply:
That's actually not why I'm doing `tzstr[3:4]`. In Python 3, the somewhat foolish decision was made that `b'x'[0] != b'x'`. The result if indexing `bytes` is an integer for some insane reason, not a single-character `bytes` or some other thing. If you take a single-character slice, it gives you `b'x'` in both Python 2 and Python 3.

Reply:
Oh, wow.  Yikes.